File: examples/python/gateway_query.py
from __future__ import annotations

from sipp import (
    CacheRuntimeConfig,
    SippClient,
    SippTextOptions,
    ContextRuntimeConfig,
    GatewayDescriptor,
    LocalModelDescriptor,
    LocalTextOptions,
    ModelPlacementConfig,
    NativeRuntimeConfig,
    ObservabilityRuntimeConfig,
    ResidencyRuntimeConfig,
    SamplingRuntimeConfig,
    SchedulerRuntimeConfig,
    set_llama_log_quiet,
)

from _support import (
    DEFAULT_CONTEXT,
    DEFAULT_MAX_TOKENS,
    DEFAULT_SEED,
    DEFAULT_TEMPERATURE,
    DEFAULT_TOP_P,
    float_env,
    gpu_layers,
    int_env,
    print_text,
    read_gateway_args,
    required_env,
)


def runtime_config(*, embeddings: bool) -> NativeRuntimeConfig:
    return NativeRuntimeConfig(
        placement=ModelPlacementConfig(gpu_layers=gpu_layers()),
        context=ContextRuntimeConfig(
            n_ctx=int_env("SIPP_CONTEXT", DEFAULT_CONTEXT),
            n_threads=int_env("SIPP_THREADS"),
            n_threads_batch=int_env("SIPP_THREADS"),
            embeddings=embeddings,
        ),
        sampling=SamplingRuntimeConfig(
            temperature=float_env("SIPP_TEMPERATURE", DEFAULT_TEMPERATURE),
            seed=int_env("SIPP_SEED", DEFAULT_SEED),
        ),
        scheduler=SchedulerRuntimeConfig(
            continuous_batching=True,
            prefill_chunk_size=0,
        ),
        cache=CacheRuntimeConfig(mode="live_slot_prefix"),
        residency=ResidencyRuntimeConfig(max_gpu_models_per_device=1),
        observability=ObservabilityRuntimeConfig(runtime_metrics=True),
    )


def text_options() -> SippTextOptions:
    return SippTextOptions(
        max_tokens=int_env("SIPP_MAX_TOKENS", DEFAULT_MAX_TOKENS),
        temperature=float_env("SIPP_TEMPERATURE", DEFAULT_TEMPERATURE),
        top_p=float_env("SIPP_TOP_P", DEFAULT_TOP_P),
    )


def main() -> None:
    model, target, prompt = read_gateway_args(
        "gateway_query", "Write one sentence about gateway inference."
    )
    set_llama_log_quiet(True)

    client = SippClient()
    local_endpoint = client.add(
        "local",
        LocalModelDescriptor(model, runtime_config(embeddings=False)),
    )
    gateway_endpoint = client.add(
        "gateway",
        GatewayDescriptor(
            target,
            required_env("SIPP_GATEWAY_URL"),
            authentication_kind="bearer",
            authentication_value=required_env("SIPP_GATEWAY_TOKEN"),
        )
    )

    # The app only needs the gateway URL, gateway bearer token, and public target.
    # Provider credentials or local model paths stay in the gateway process.
    local = client.query(
        prompt,
        endpoint=local_endpoint,
        options=text_options(),
        local=LocalTextOptions(context_key="python-gateway-query-local"),
    ).result()
    gateway = client.query(
        prompt,
        endpoint=gateway_endpoint,
        options=text_options(),
    ).result()

    print("local:")
    print_text(local)
    print("gateway:")
    print_text(gateway)


if __name__ == "__main__":
    main()

File: examples/rust/src/gateway_query.rs
mod support;

use std::path::PathBuf;

use cogentlm::backend::set_llama_log_quiet;
use cogentlm::engine::{
    CacheRuntimeConfig, ContextRuntimeConfig, GpuLayerConfig, KvReuseMode, ModelPlacementConfig,
    NativeRuntimeConfig, ObservabilityRuntimeConfig, ResidencyRuntimeConfig, SamplingRuntimeConfig,
    SchedulerRuntimeConfig,
};
use cogentlm::{
    CogentClient, CogentQueryRequest, CogentTextOptions, EndpointDescriptor, GatewayAuthentication,
    GatewayEndpointConfig, GatewayRoutes, GatewaySecret, GatewayTimeoutPolicy, LocalTextOptions,
};
use futures::executor::block_on;

fn main() -> support::ExampleResult<()> {
    block_on(async {
        let args = support::gateway_args(
            "Write one sentence about gateway inference.",
            "gateway_query",
        )?;
        set_llama_log_quiet(true);

        let mut client = CogentClient::new();
        let local_endpoint = client
            .add(
                "local",
                EndpointDescriptor::local(args.model_path, runtime_config(false, None)),
            )
            .await?;
        let config = GatewayEndpointConfig {
            target: args.target.clone(),
            base_url: support::required_env("COGENTLM_GATEWAY_URL")?,
            routes: GatewayRoutes::default(),
            authentication: GatewayAuthentication::Bearer(GatewaySecret::new(
                support::required_env("COGENTLM_GATEWAY_TOKEN")?,
            )),
            static_headers: Default::default(),
            timeouts: GatewayTimeoutPolicy::default(),
            protocol_options: Default::default(),
        };
        let gateway_endpoint = client
            .add("gateway", EndpointDescriptor::gateway(config))
            .await?;

        let local = client
            .query(CogentQueryRequest {
                endpoint: Some(local_endpoint),
                prompt: args.input.clone(),
                options: text_options(),
                local: LocalTextOptions {
                    context_key: Some("rust-gateway-query-local".to_string()),
                    ..Default::default()
                },
                ..Default::default()
            })
            .await?;

        let gateway = client
            .query(CogentQueryRequest {
                endpoint: Some(gateway_endpoint),
                prompt: args.input,
                options: text_options(),
                ..Default::default()
            })
            .await?;

        println!("local:");
        support::print_text(local);
        println!("gateway:");
        support::print_text(gateway);
        Ok(())
    })
}

fn runtime_config(embeddings: bool, projector_path: Option<PathBuf>) -> NativeRuntimeConfig {
    NativeRuntimeConfig {
        placement: ModelPlacementConfig {
            gpu_layers: support::env_parse("COGENTLM_GPU_LAYERS")
                .map(GpuLayerConfig::from_layer_count)
                .unwrap_or(GpuLayerConfig::Auto),
            ..Default::default()
        },
        context: ContextRuntimeConfig {
            n_ctx: support::env_parse("COGENTLM_CONTEXT").or(Some(support::DEFAULT_CONTEXT)),
            n_threads: support::env_parse("COGENTLM_THREADS"),
            n_threads_batch: support::env_parse("COGENTLM_THREADS"),
            embeddings: embeddings.then_some(true),
            ..Default::default()
        },
        sampling: SamplingRuntimeConfig {
            temperature: support::env_parse("COGENTLM_TEMPERATURE")
                .or(Some(support::DEFAULT_TEMPERATURE)),
            seed: support::env_parse("COGENTLM_SEED").or(Some(support::DEFAULT_SEED)),
            ..Default::default()
        },
        scheduler: SchedulerRuntimeConfig {
            continuous_batching: true,
            prefill_chunk_size: 0,
            ..Default::default()
        },
        cache: CacheRuntimeConfig {
            mode: KvReuseMode::LiveSlotPrefix,
            ..Default::default()
        },
        multimodal: cogentlm::engine::MultimodalRuntimeConfig {
            projector_path: projector_path.map(|path| path.to_string_lossy().into_owned()),
            ..Default::default()
        },
        residency: ResidencyRuntimeConfig {
            max_gpu_models_per_device: 1,
            ..Default::default()
        },
        observability: ObservabilityRuntimeConfig {
            runtime_metrics: true,
            backend_profiling: false,
        },
    }
}

fn text_options() -> CogentTextOptions {
    CogentTextOptions {
        max_tokens: support::env_parse("COGENTLM_MAX_TOKENS").or(Some(support::DEFAULT_MAX_TOKENS)),
        temperature: support::env_parse("COGENTLM_TEMPERATURE")
            .or(Some(support::DEFAULT_TEMPERATURE)),
        top_p: support::env_parse("COGENTLM_TOP_P").or(Some(support::DEFAULT_TOP_P)),
        stop: Vec::new(),
    }
}
